DeepSpeech.ipynb

```diff
@@ -1320,170 +1292,390 @@
+    "    # Creating the execution context\n",
+    "    execution_context = create_execution_context(set_name)\n",
     "    \n",
-    "    # Training results parameters\n",
-    "    train_results_params = (tower_labels, tower_decodings, tower_distances, tower_total_losses)\n",
+    "    # Starting the execution context \n",
+    "    session = start_execution_context(execution_context, model_path=model_path)\n",
     "    \n",
-    "    # Prepare tensor board logging\n",
-    "    merged = tf.merge_all_summaries()\n",
-    "    writer = tf.train.SummaryWriter(log_dir, session.graph)\n",
+    "    # Applying the batches\n",
+    "    batch_set_result = apply_batches(execution_context, session, query_report=query_report)\n",
     "    \n",
-    "    # Train parameters\n",
-    "    train_params = [apply_gradient_op, avg_loss, merged]\n",
+    "    # Cleaning up\n",
+    "    stop_execution_context(execution_context, session)\n",
+    "\n",
+    "    # Returning batch_set_result from apply_batches\n",
+    "    return batch_set_result"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Training"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now, as we have prepared all the apropos operators and methods, we can create the method which trains the network."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "def train():\n",
+    "\n",
+    "    # Creating the training execution context\n",
+    "    train_context = create_execution_context('train')\n",
     "\n",
     "    # Init recent word error rate levels\n",
     "    train_wer = 0.0\n",
     "    dev_wer = 0.0\n",
-    "    \n",
+    "\n",
+    "    hibernation_path = None\n",
+    "\n",
     "    # Possibly restore checkpoint\n",
     "    start_epoch = 0\n",
     "    if restore_checkpoint:\n",
     "        checkpoint = tf.train.get_checkpoint_state(checkpoint_dir)\n",
     "        if checkpoint and checkpoint.model_checkpoint_path:\n",
-    "            saver.restore(session, checkpoint.model_checkpoint_path)\n",
+    "            hibernation_path = checkpoint.model_checkpoint_path\n",
     "            start_epoch = int(checkpoint.model_checkpoint_path.split('-')[-1])\n",
     "            print 'Resuming training from epoch %d' % (start_epoch + 1)\n",
     "    \n",
     "    # Loop over the data set for training_epochs epochs\n",
     "    for epoch in range(start_epoch, training_iters):\n",
+    "\n",
+    "        print \"STARTING Epoch\", '%04d' % (epoch)\n",
     "        \n",
-    "        print \"\\n\\n\\nSTARTING Epoch\", '%04d' % (epoch)\n",
+    "        if epoch == 0 or hibernation_path is not None:\n",
+    "            if hibernation_path is not None:\n",
+    "                print \"Resuming training session from\", \"%s\" % hibernation_path, \"...\"\n",
```

Comment:
But some executions of `start_execution_context()` print out a message. Some don't.